### PR TITLE
Update jupyterhub config to enable hostIPC access for select users

### PR DESCRIPTION
**Pull request by maulikjs**

### CODE REVIEW

**Comment by maulikjs** [comment]
@LaVLaS Can you give this MR a look?

**Comment by anishasthana** [comment]
Landon is on PTO, so putting it on /hold until he's back

**Comment by maulikjs** [comment]
One thing to note here is that we will have to ask PSI to update the restricted SCC
`allowHostIPC: false -> true`

File: jupyterhub/bases/custom-profiles/jupyterhub-singleuser-profiles.yaml [hunk 1]
```diff
@@ -76,6 +76,9 @@ profiles:
   resources:
     mem_limit: 256Gi
     cpu_limit: 16
+  env:
+    enable_host_ipc: true
```
Comment (LaVLaS) [inline jupyterhub/bases/custom-profiles/jupyterhub-singleuser-profiles.yaml]: Should this be all caps `ENABLE_HOST_IPC` as a standard?

File: jupyterhub/bases/jupyterhub/configs/jupyterhub_config.py [hunk 2]
```diff
@@ -1,11 +1,28 @@
...
   if spawner.gpu_mode and spawner.gpu_mode == "selinux" and \
        spawner.extra_resource_limits and "nvidia.com/gpu" in spawner.extra_resource_limits:
     # Currently a bug in RHEL Docker 1.13 whereby /dev IPC dirs get inconsistent MCS
     pod.spec.security_context.se_linux_options = V1SELinuxOptions(type='nvidia_container_t',level='s0')
+
+  for item in nb_container_env:
+    if item.name == "enable_host_ipc":
+      # This serviceAccount should have the same permission as the 'default' serviceAccount
+      #   with an applied SCC for hostIPC access
+      pod.spec.service_account = 'jupyterhub-nb-hostipc-access'
```
Comment (LaVLaS) [inline jupyterhub/bases/jupyterhub/configs/jupyterhub_config.py]: This wasn't required when we tested using the okd environment so you'll want to confirm that it is required for PSI

File: jupyterhub/bases/jupyterhub/configs/jupyterhub_config.py [hunk 3]
```diff
@@ -1,11 +1,28 @@
...
+  """
+  This method is responsible for modifying the final notebook pod spec before submitting it to the k8s api
+  Modifying the pod based on the V1Pod and V1PodSpec avaialable in the kubernetes python API
+  https://github.com/kubernetes-client/python/blob/master/kubernetes/docs
+  """
+  # Pod container zero is the 'notebook' container with the env var we need to check
+  nb_container_env = pod.spec.containers[0].env
+
   # Apply profile from singleuser-profiles
   apply_pod_profile(spawner, pod)
```
Comment (LaVLaS) [inline jupyterhub/bases/jupyterhub/configs/jupyterhub_config.py]: I forgot to move this to the first statement executed in the function as standard practice for calling the parent method AND the pod object will be modified by `apply_pod_profile` so we want to reference it after the parent modifies it

**Review by LaVLaS — COMMENTED** [review]
Few comments. I'm happy to review this while I'm on PTO since it was delayed for so long...by me

File: jupyterhub/bases/jupyterhub/configs/jupyterhub_config.py (hunk 2 above)
Comment (maulikjs) [reply jupyterhub/bases/jupyterhub/configs/jupyterhub_config.py]: I have commented this block out for now, we can remove it in a subsequent PR if it works in prod.

**Comment by maulikjs** [comment]
@LaVLaS can you recheck?

File: jupyterhub/bases/jupyterhub/configs/jupyterhub_config.py [hunk 4]
```diff
@@ -1,11 +1,31 @@
...
+  https://github.com/kubernetes-client/python/blob/master/kubernetes/docs
+  """
+  # Apply profile from singleuser-profiles
+  apply_pod_profile(spawner, pod)
+
+  # Pod container zero is the 'notebook' container with the env var we need to check
+  nb_container_env = pod.spec.containers[0].env
+
   # Apply profile from singleuser-profiles
   apply_pod_profile(spawner, pod)
```
Comment (LaVLaS) [inline jupyterhub/bases/jupyterhub/configs/jupyterhub_config.py]: This is a dupe

**Comment by LaVLaS** [comment]
/lgtm

I only have one OPTIONAL comment, do you want to remove the creation/deployment of the `jupyterhub-nb-hostipc-access` service account since it is commented out in `idh_apply_pod_profile()` and will never be used unless required for deployment internally?

**Pull request merged**